pallets/block-rewards/src/lib.rs

```diff
@@ -0,0 +1,475 @@
+			}
+		}
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+		fn build(&self) {
+			T::Rewards::attach_currency((T::Domain::get(), STAKE_CURRENCY_ID), COLLATOR_GROUP_ID)
+				.map_err(|e| log::error!("Failed to attach currency to collator group: {:?}", e))
+				.ok();
+
+			ActiveSessionData::<T>::mutate(|session_data| {
+				session_data.collator_count = self.collators.len().saturated_into();
+				session_data.collator_reward = self.collator_reward;
+				session_data.total_reward = self.total_reward;
+			});
+
+			// Enables rewards already in genesis session.
+			for collator in &self.collators {
+				Pallet::<T>::do_init_collator(collator)
+					.map_err(|e| {
+						log::error!("Failed to init genesis collators for rewards: {:?}", e);
+					})
+					.ok();
+			}
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Claims the reward the associated to a currency.
+		/// The reward will be transferred to the target account.
+		#[pallet::weight(T::WeightInfo::claim_reward())]
+		#[transactional]
+		pub fn claim_reward(origin: OriginFor<T>, account_id: T::AccountId) -> DispatchResult {
+			ensure_signed(origin)?;
+
+			T::Rewards::claim_reward((T::Domain::get(), STAKE_CURRENCY_ID), &account_id).map(|_| ())
+		}
+
+		/// Admin method to set the reward amount for a collator used for the next sessions.
+		/// Current session is not affected by this call.
+		#[pallet::weight(T::WeightInfo::set_collator_reward())]
+		pub fn set_collator_reward(
+			origin: OriginFor<T>,
+			collator_reward_per_session: T::Balance,
+		) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			NextSessionChanges::<T>::mutate(|changes| {
+				changes.collator_reward = Some(collator_reward_per_session);
+			});
+
+			Ok(())
+		}
+
+		/// Admin method to set the total reward distribution for the next sessions.
+		/// Current session is not affected by this call.
+		///
+		/// Throws if total_reward < collator_reward * collator_count.
+		#[pallet::weight(T::WeightInfo::set_total_reward())]
+		pub fn set_total_reward(
+			origin: OriginFor<T>,
+			total_reward_per_session: T::Balance,
+		) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			NextSessionChanges::<T>::try_mutate(|changes| {
+				let current = ActiveSessionData::<T>::get();
+				let total_collator_rewards = changes
+					.collator_reward
+					.unwrap_or(current.collator_reward)
+					.saturating_mul(
+						changes
+							.collator_count
+							.unwrap_or(current.collator_count)
+							.into(),
+					);
+				ensure!(
+					total_reward_per_session >= total_collator_rewards,
+					Error::<T>::InsufficientTotalReward
+				);
+
+				changes.total_reward = Some(total_reward_per_session);
+				Ok(())
+			})
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	/// Mint default amount of stake for target address and deposit stake.
+	/// Enables receiving rewards onwards.
+	///
+	/// Weight (6 reads, 6 writes):
```

Comment:
For now, knowing the correct weight would only be necessary for the migration. We have 8 collators each on Altair and Centrifuge. So we should be fine without an accurate benchmark IMO. WDYT?